### ISSUE TITLE
Frictionless => CKAN: Refactor and improve code and tests

### ISSUE THREAD

**Issue by sglavoie** [frictionlessdata/frictionless-ckan-mapper#21]
We want to refactor and improve Frictionless => CKAN code, e.g.

## Acceptance

* [x] Extras handling including parsing of JSON values, e.g. schema.
* [x] Full set of mappings.

Approach: _no_ changes wherever possible so we support roundtripping.

## Tasks

**DONE**: Everything ticked as done with no explanation next to it is being addressed in the PR #30.

Design algorithm => incrementally implement based on tests

* [x] Design algorithm and document in docstring -- **DONE**: See [comment below](https://github.com/frictionlessdata/frictionless-ckan-mapper/issues/21#issuecomment-647687833).
* [x] Migrate existing tests and have them pass
	* [x] list existing tests -- **DONE**: See Analysis section below.
* [x] Refactor existing tests
	* [x] Dedupe simple pass through tests e.g. for Resource description, hash etc (which just pass through)
* [x] Resource tests
* [x] Package tests based on our detailed mapping analysis in spreadsheet https://docs.google.com/spreadsheets/d/1XdqGTFni5Jfs8AMbcbfsP7m11h9mOHS0eDtUZtqGVSg/edit#gid=2099009729 TODO
  * [x] `path_is_mapped_to_url` - even if relative POSIX path!
  - [x] test_author_and_maintainer_is_converted
  - [x] test_extras_is_converted - i.e. all keys not in the accepted keys on https://docs.ckan.org/en/2.8/api/index.html#ckan.logic.action.create.package_create are mapped into extra with values jsonified ...
  - [x] test_license_is_converted
  * [x] Clarify current doubts regarding the FD => CKAN mappings. -- **DONE**: Asking questions in the [comment below](https://github.com/frictionlessdata/frictionless-ckan-mapper/issues/21#issuecomment-647687833).

### Existing tests

```
Data package

test_basic_datapackage_in_setup_is_valid        # probably not needed
test_datapackage_only_requires_some_fields_to_be_valid
test_datapackage_name_title_and_version
test_name_is_lowercased
test_datapackage_description
test_datapackage_license_as_string
test_datapackage_license_as_unicode    # does not seem too useful, skipping for now
test_datapackage_license_as_dict
test_datapackage_sources
test_datapackage_author_as_string
test_datapackage_author_as_unicode
test_datapackage_author_as_string_without_email
test_datapackage_author_as_dict
test_datapackage_keywords
test_datapackage_extras

Resource

test_resource_name_is_used_if_theres_no_title
test_resource_title_is_used_as_name
test_resource_url
test_resource_url_is_set_to_its_remote_data_path
test_resource_description
test_resource_format
test_resource_hash
test_resource_schema
test_resource_path_is_set_to_its_local_data_path
```

### Additional tests to consider

* path_is_mapped_to_url
  * even if relative POSIX path!

```
test_author_is_converted
test_extras_is_converted
test_license_is_converted
```

### Values get JSONified by CKAN

I just tried that for an existing package I had and it got JSONified.

I sent this:
```
curl -X POST https://demo.ckan.org/api/3/action/resource_create -H "Authorization: CORRECT-KEY-HERE" -d '{    7 ↵
  "package_id": "ckan-to-frictionless-conversion",
  "url":  "https://raw.githubusercontent.com/frictionlessdata/test-data/master/files/csv/100kb.csv",
  "description": "This is the best resource ever!" ,
  "name": "brand-new-resource",
  "more": {"key": "not jsonified"}
}'
```
And received the response:
```
{
   "help":"https://demo.ckan.org/api/3/action/help_show?name=resource_create",
   "success":true,
   "result":{
      "cache_last_updated":null,
      "cache_url":null,
      "mimetype_inner":null,
      "hash":"",
      "description":"This is the best resource ever!",
      "format":"CSV",
      "url":"https://raw.githubusercontent.com/frictionlessdata/test-data/master/files/csv/100kb.csv",
      "created":"2020-06-25T14:10:29.264869",
      "state":"active",
      "name":"brand-new-resource",
      "package_id":"99575b35-8a88-4fd9-b0dc-b9d0479c9b2c",
      "last_modified":null,
      "mimetype":null,
      "url_type":null,
      "position":2,
      "revision_id":"19f22475-ff52-46aa-8918-627e3359dc31",
      "size":null,
      "datastore_active":false,
      "id":"331a9e6e-2230-4bd8-84a4-3e2b66682f98",
      "resource_type":null,
      "more":"{'key': 'not jsonified'}"
   }
}
```

If you search for the keyword `more` you can see it here: https://demo.ckan.org/api/3/action/package_show?id=ckan-to-frictionless-conversion

**Comment by sglavoie** [comment]
@rufuspollock, I created a new sheet in the document we were using: [Frictionless => CKAN](https://docs.google.com/spreadsheets/d/1XdqGTFni5Jfs8AMbcbfsP7m11h9mOHS0eDtUZtqGVSg/edit#gid=1525638854).

After going through those mappings in order to make progress on the task "Design algorithm and document in docstring", I still have some questions. I pushed (no PR yet) the basic structure we will need to test the conversion FD => CKAN here: https://github.com/frictionlessdata/frictionless-ckan-mapper/tree/tests/fd-to-ckan

This is the description of the commit:

[fd-to-ckan][s]: Lay down the structural foundation for tests

* Create `frictionless_to_ckan.py` with basic class and functions needed.
* Add comments in `frictionless_to_ckan.py` and document with a
  docstring.
* Create the skeleton for `test_frictionless_to_ckan.py`.

---

### Questions

Here are the questions I left as comments/TODOs in the code:

On [lines 11-13](https://github.com/frictionlessdata/frictionless-ckan-mapper/blob/392cf7352ca66798b117291c4e4143b6bd6e5da8/frictionless_ckan_mapper/frictionless_to_ckan.py#L11-L13) in `frictionless_to_ckan.py`:

* How do we map the required CKAN `package_id`?
* In CKAN, `private` is another required field: set the default value to False?

On [lines 21-26](https://github.com/frictionlessdata/frictionless-ckan-mapper/blob/392cf7352ca66798b117291c4e4143b6bd6e5da8/frictionless_ckan_mapper/frictionless_to_ckan.py#L21-L26) in `frictionless_to_ckan.py`:

* Do we passthrough all the FD properties not found in CKAN?
  * `data`: This replaces Resource.path if it is not specified. However, it is not an URL. Should it just passthrough?
  * `homepage`: Does this passthrough? What if we also have FD Resource.path => we can't map both to CKAN Resource.url
  * `licenses`: How should we map multiple licenses from Frictionless to CKAN?

---

### Additional info about the commit

The documentation in the docstring for a **FD=>CKAN resource conversion** can be found [on lines 36-40](https://github.com/frictionlessdata/frictionless-ckan-mapper/blob/392cf7352ca66798b117291c4e4143b6bd6e5da8/frictionless_ckan_mapper/frictionless_to_ckan.py#L36-L40):
```
1. Remove unneeded keys (are there any keys to remove?)
2. Convert extras: "dict" to "list of dict".
3. Map keys from Frictionless to CKAN (and reformat if needed).
4. Remove keys with null values.
5. Apply special formatting (if any) for key fields e.g. slugify
```

The documentation in the docstring for a **FD=>CKAN package conversion** can be found [on lines 49-53](https://github.com/frictionlessdata/frictionless-ckan-mapper/blob/392cf7352ca66798b117291c4e4143b6bd6e5da8/frictionless_ckan_mapper/frictionless_to_ckan.py#L49-L53):
```
1. Copy extras across.  # Is this true?
2. Map keys from Frictionless to CKAN (and reformat if needed).
3. Remove keys with null values (CKAN has a lot of null valued keys).
4. Remove unneeded keys.
5. Apply special formatting for key fields.
```

The **skeleton for testing FD=>CKAN conversions** [can be found here](https://github.com/frictionlessdata/frictionless-ckan-mapper/blob/392cf7352ca66798b117291c4e4143b6bd6e5da8/tests/test_frictionless_to_ckan.py).

**Comment by rufuspollock** [comment]
@sglavoie great and ... it's a little bit tough to process that much info 😉  I'd recommend doing PRs more and doing exec summary of key questions / blockers and your hypothesis for answers http://playbook.datopian.com/hypotheses

**Comment by sglavoie** [comment]
**TL;DR**: The PR #30 adds all tests discussed in this issue, implement them and refactor classes into simple functions as discussed.

I know this is a relatively big PR, but there are 19 commits separating concerns very clearly, so I hope it's good to go (plus, Travis didn't complain for neither Python 2 nor Python 3, which is great!). Also, this is to facilitate merging into master assuming mostly everything should be as expected or at least more or less easy to revert on a commit per commit basis.

I will go through the Playbook (again). :+1:

**Comment by sglavoie** [comment]
**FIXED**: Closing as the last couple of commits addressed the final tasks for this issue.

* https://github.com/frictionlessdata/frictionless-ckan-mapper/commit/0bb722d0f62a8d3d57501c2e2a7630390309bb23
* https://github.com/frictionlessdata/frictionless-ckan-mapper/commit/b97818b9cd05f5a848ef9b1de12943a2ab376150 (small fix for https://github.com/frictionlessdata/frictionless-ckan-mapper/commit/0bb722d0f62a8d3d57501c2e2a7630390309bb23)
